File: Student-Performance-Predictor/frontend/js/prediction.js
let predPage = 1;

document.addEventListener('DOMContentLoaded', () => {
    if (window.location.pathname.includes('prediction.html')) {
        loadPredictionHistory();

        document.getElementById('predictionForm').addEventListener('submit', async (e) => {
            e.preventDefault();
            const btn = e.target.querySelector('button[type="submit"]');
            btn.disabled = true;
            btn.innerHTML = '<span class="spinner"></span> Predicting...';

            try {
                const payload = {
                    student_id: document.getElementById('predStudentId').value.trim(),
                    student_name: document.getElementById('predStudentName').value.trim(),
                    department: document.getElementById('predDept').value,
                    semester: parseInt(document.getElementById('predSem').value),
                    attendance: parseFloat(document.getElementById('predAttendance').value) || 0,
                    study_hours: parseFloat(document.getElementById('predStudy').value) || 0,
                    assignment_score: parseFloat(document.getElementById('predAssign').value) || 0,
                    internal_marks: parseFloat(document.getElementById('predInternal').value) || 0,
                    previous_marks: parseFloat(document.getElementById('predPrevious').value) || 0,
                    algorithm: document.getElementById('predAlgo').value,
                };

                const res = await fetch(`${API_URL}/api/predictions/predict`, {
                    method: 'POST',
                    headers: { 'Content-Type': 'application/json' },
                    credentials: 'include',
                    body: JSON.stringify(payload)
                });
                const data = await res.json();

                if (!res.ok) { showToast(data.error || 'Prediction failed', 'error'); return; }

                const p = data.prediction;
                const container = document.getElementById('resultContainer');
                const resultDiv = document.getElementById('predictionResult');
                container.style.display = 'block';

                let recHtml = p.recommendation ? p.recommendation.split(' ').slice(0, 20).join(' ') : 'Continue current habits.';
                if (recHtml.length > 100) recHtml = recHtml.substring(0, 100) + '...';

                resultDiv.innerHTML = `
                    <div class="prediction-score mb-2">${p.predicted_marks}</div>
                    <div class="mb-3">${getGradeBadge(p.grade)} ${getRiskBadge(p.risk_level)}</div>
                    <div class="mb-2">${getPerformanceBadge(p.performance)}</div>
                    <hr>
                    <p class="small text-muted mb-0">${recHtml}</p>
                `;

                // Show insights
                const insightsContainer = document.getElementById('insightsContainer');
                if (data.insights && data.insights.length > 0) {
                    insightsContainer.innerHTML = data.insights.map(i =>
                        `<div class="alert alert-warning py-2"><i class="fas fa-lightbulb me-2"></i>${i}</div>`
                    ).join('');
                } else {
                    insightsContainer.innerHTML = '';
                }

                showToast('Prediction completed!', 'success');
                loadPredictionHistory();

            } catch (e) {
                showToast('Prediction failed. Check if model is trained.', 'error');
            } finally {
                btn.disabled = false;
                btn.innerHTML = '<i class="fas fa-calculator me-2"></i> Predict Performance';
            }
        });
    }
});

async function loadPredictionHistory(page = predPage) {
    predPage = page;
    const search = document.getElementById('predHistSearch')?.value || '';
    const dept = document.getElementById('filterDepartment')?.value || '';
    const perf = document.getElementById('filterPerformance')?.value || '';
    try {
        const url = `${API_URL}/api/predictions/history?page=${page}&per_page=20&search=${search}&department=${dept}&performance=${perf}`;
        const res = await fetch(url, { credentials: 'include' });
        const data = await res.json();
        const tbody = document.getElementById('predHistoryBody');
        const pagination = document.getElementById('predPagination');
        if (!tbody) return;
        if (!data.predictions || data.predictions.length === 0) {
            tbody.innerHTML = '<tr><td colspan="8" class="text-center text-muted py-4">No predictions found</td></tr>';
            if (pagination) pagination.innerHTML = '';
            return;
        }
        tbody.innerHTML = data.predictions.map(p => `
            <tr>
                <td><strong>${p.student_name || 'N/A'}</strong><br><small class="text-muted">${p.student_id || ''}</small></td>
                <td>${p.department || '-'}</td>
                <td><strong>${p.predicted_marks || 0}</strong></td>
                <td>${getGradeBadge(p.grade)}</td>
                <td>${getPerformanceBadge(p.performance)}</td>
                <td>${getRiskBadge(p.risk_level)}</td>
                <td>${formatDate(p.created_at)}</td>
                <td class="actions">
                    <button class="btn-delete" onclick="deletePrediction('${p.id}')" title="Delete"><i class="fas fa-trash"></i></button>
                </td>
            </tr>
        `).join('');
        if (pagination) {
            paginate(data.total_pages || 1, data.page, 'predPagination', 'loadPredictionHistory');
        }
    } catch { showToast('Failed to load history', 'error'); }
}

async function deletePrediction(id) {
    if (!confirm('Delete this prediction?')) return;
    try {
        await fetch(`${API_URL}/api/predictions/${id}`, { method: 'DELETE', credentials: 'include' });
        showToast('Prediction deleted', 'success');
        loadPredictionHistory();
    } catch { showToast('Failed to delete', 'error'); }
}

async function exportPredictions() {
    try {
        const res = await fetch(`${API_URL}/api/predictions/export`, { credentials: 'include' });
        const data = await res.json();
        if (!data.predictions) { showToast('No predictions to export', 'warning'); return; }
        let csv = 'Student,Department,Grade,Predicted Marks,Performance,Risk Level,Date\n';
        data.predictions.forEach(p => {
            csv += `"${p.student_name || ''}","${p.department || ''}","${p.grade || ''}",${p.predicted_marks || 0},"${p.performance || ''}","${p.risk_level || ''}","${p.created_at || ''}"\n`;
        });
        const blob = new Blob([csv], { type: 'text/csv' });
        const url = URL.createObjectURL(blob);
        const a = document.createElement('a');
        a.href = url;
        a.download = `predictions_${new Date().toISOString().slice(0, 10)}.csv`;
        a.click();
        URL.revokeObjectURL(url);
        showToast('Predictions exported', 'success');
    } catch { showToast('Export failed', 'error'); }
}

function paginate(totalPages, current, containerId, callback) {
    const container = document.getElementById(containerId);
    if (!container) return;
    container.innerHTML = '';
    if (totalPages <= 1) return;
    container.innerHTML += `<button onclick="${callback}(1)" ${current === 1 ? 'disabled' : ''}><i class="fas fa-chevron-left"></i></button>`;
    for (let i = Math.max(1, current - 2); i <= Math.min(totalPages, current + 2); i++) {
        container.innerHTML += `<button class="${i === current ? 'active' : ''}" onclick="${callback}(${i})">${i}</button>`;
    }
    container.innerHTML += `<button onclick="${callback}(${totalPages})" ${current === totalPages ? 'disabled' : ''}><i class="fas fa-chevron-right"></i></button>`;
}

File: Student-Performance-Predictor/backend/routes/prediction.py
from flask import Blueprint, request, jsonify
from database.db import _insert, _search, _delete, _all, _find
from routes.auth import login_required
from services.ml_service import load_model, predict_performance, get_ai_insights

prediction_bp = Blueprint("prediction", __name__, url_prefix="/api/predictions")


@prediction_bp.route("/predict", methods=["POST"])
@login_required
def predict():
    try:
        data = request.get_json()
        if not data:
            return jsonify({"error": "No data provided"}), 400
        model = load_model()
        if model is None:
            return jsonify({"error": "No trained model found. Please train a model first."}), 400
        features = {"attendance": float(data.get("attendance", 0)), "study_hours": float(data.get("study_hours", 0)), "assignment_score": float(data.get("assignment_score", 0)), "internal_marks": float(data.get("internal_marks", 0)), "previous_marks": float(data.get("previous_marks", 0))}
        result = predict_performance(model, features)
        insights = get_ai_insights(features)
        rec = _insert("predictions", {"student_id": data.get("student_id", "N/A"), "student_name": data.get("student_name", ""), "department": data.get("department", ""), "semester": int(data.get("semester", 1)), "attendance": features["attendance"], "study_hours": features["study_hours"], "assignment_score": features["assignment_score"], "internal_marks": features["internal_marks"], "previous_marks": features["previous_marks"], "predicted_marks": result["predicted_marks"], "grade": result["grade"], "performance": result["performance"], "risk_level": result["risk_level"], "recommendation": result["recommendation"], "algorithm": data.get("algorithm", "Random Forest"), "accuracy": float(data.get("accuracy", 0))})
        return jsonify({"prediction": result, "insights": insights, "record_id": rec.data[0]["id"] if rec.data else None}), 200
    except Exception as e:
        return jsonify({"error": str(e)}), 500


@prediction_bp.route("/history", methods=["GET"])
@login_required
def get_history():
    try:
        page = request.args.get("page", 1, type=int)
        per_page = request.args.get("per_page", 20, type=int)
        search = request.args.get("search", "").strip()
        dept = request.args.get("department", "").strip()
        perf = request.args.get("performance", "").strip()
        filters = {}
        if dept: filters["department"] = dept
        if perf: filters["performance"] = perf
        result = _search("predictions", search, ["student_name", "student_id"], page=page, per_page=per_page, filters=filters)
        return jsonify({"predictions": result.data, "total": result.count, "page": page, "per_page": per_page, "total_pages": max(1, -(-result.count // per_page))}), 200
    except Exception as e:
        return jsonify({"error": str(e)}), 500


@prediction_bp.route("/<prediction_id>", methods=["DELETE"])
@login_required
def delete_prediction(prediction_id):
    try:
        _delete("predictions", "id", prediction_id)
        return jsonify({"message": "Prediction deleted successfully"}), 200
    except Exception as e:
        return jsonify({"error": str(e)}), 500


@prediction_bp.route("/export", methods=["GET"])
@login_required
def export_predictions():
    try:
        result = _all("predictions", order="created_at")
        return jsonify({"predictions": result.data}), 200
    except Exception as e:
        return jsonify({"error": str(e)}), 500